### ISSUE TITLE
Question: Can tslog format logs from a tranport?

### ISSUE THREAD

**Issue by LiamLamb** [fullstack-build/tslog#205]
Hi 👋,

I am using your logging tool to develop an interval vscode extension in my organisation.
The extension has multiple output channels for each type of HTTP client talking to different internal services. 

What I like is that I can logically split my logs across each client's output channel, making extension logging organised, so I can look at client logs in isolation based on their client type, providing me with a developer experience akin to Serilog in .NET. 

When running the extension in vscode, as a user looking at the logs in the output channel, I wouldn't need to see the structured log object, but rather a nicely formatted message (although the structured log itself would be stored to file/pushed to a remote server) in the transport callback. 

Here's a simplified shot of what I am doing:

```ts
// loggerBuilder.ts
transport(func: (payload: ILogObj & ILogObjMeta) => void): ILoggerBuilder {
  this._transports.push(func);
  
  return this;
}

build() {
   ...
   this._transports.forEach((t) => logger.attachTransport(t));
   ...
}

// client
...
loggerBuilder
  .transport((log) => {
      // Log to file
  } )
  .transport((log) => {
    // Logging as structured log for now, but ideally formatted here
    outputChannel.appendLine(JSON.stringify(logObj));
  });

...
```

It would be convenient to be able to format the JSON object from the transport, in some way. 

Could you provide some insight into whether this is an already-supported feature and if I am missing something?

Thanks 🙂

**Comment by terehov** [comment]
It sounds like you're suggesting the ability to choose between a `json` or `pretty` format for logs, depending on the attached transport, correct? Alternatively to expose a helper that would transform an `ILogObj` into a `pretty`string, right?

**Comment by LiamLamb** [comment]
Yes, that's correct - either approach would work. 

I can see how this might not be a common use case, as in most cases logging to console would be sufficient, but given that vscode has a notion of output channels, this would be extremely useful.

Let me know what your thoughts are

Thanks

**Comment by JustinGrote** [comment]
This is what I do as a workaround:
```ts
const log = new Logger<DefaultLog>({
	name: 'default',
	type: 'pretty',
	prettyErrorLoggerNameDelimiter: "-",
	prettyErrorParentNamesSeparator: "-",
	stylePrettyLogs: true,
	argumentsArrayName: "args",
	overwrite: {
		transportFormatted: () => { return } 		// We want pretty formatting but no default output and will attach a transport later
	}
})
```

If I do this then my attached transport still has the formatted message and I can output it accordingly.

**Comment by JustinGrote** [comment]
@LiamLamb I vote for a simple exported function that runs the "pretty" process for the current logger on any string, that's going to be the most flexible and least breaking vs. adding a property on ILogObj, etc.

**Comment by JustinGrote** [comment]
@terehov one question I have is why pretty and json are "special" and have their own special steps. Why are they not just implemented as additional transports or maybe transports that are "pre-attached" unless "hidden" is specified.

My main issue is the parameters available in the formattedTransport override is more extensive (including settings) than is available in the attachTransport param which is just the log object. I think it should be the same function signature for both, with pretty and json just being two implementations of that signature.

**Comment by James4Ever0** [comment]
How to resolve the source line of Typescript VSCode extension using `tslog`? I have a simple hack here: https://github.com/cline/cline/issues/1229#issuecomment-2589675331

**Comment by terehov** [comment]
Yes — supported in the upcoming **v5**. Each transport can declare its own `format` (`"pretty"`, `"json"`, or a custom formatter function), independent of the logger-wide `type`. So you can keep structured records going to file/remote while an output-channel transport receives a nicely formatted string. Closing as addressed in v5. Thanks!
